// meterify/dist/utils/option-formatters.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatLogQuery = exports.formatOptions = exports.formatRange = exports.fromETHBlockNumberOrHash = exports.fromETHBlockNumber = void 0;
const utils = require("./");
exports.fromETHBlockNumber = function (blockNumber) {
    if (typeof blockNumber === 'number') {
        return blockNumber;
    }
    else if (typeof blockNumber === 'string') {
        if (blockNumber === 'earliest') {
            return 0;
        }
        else if (blockNumber === 'latest' || blockNumber === 'pending') {
            return 'best';
        }
        else {
            const num = utils.toInteger(blockNumber);
            return num || num === 0 ? num : 'best';
        }
    }
    else {
        return 'best';
    }
};
exports.fromETHBlockNumberOrHash = function (blockRevision) {
    if (/^(-0x|0x)?[0-9a-fA-F]{64}$/i.test(blockRevision)) {
        return blockRevision;
    }
    else {
        return exports.fromETHBlockNumber(blockRevision);
    }
};
exports.formatRange = function (range) {
    const ret = {};
    if (range.unit !== 'block' && range.unit !== 'time') {
        return null;
    }
    else {
        ret.unit = range.unit;
    }
    if (range.hasOwnProperty('from')) {
        const temp = exports.fromETHBlockNumber(range.from);
        if (temp !== 'best') {
            ret.from = temp;
        }
        else {
            ret.from = 0;
        }
    }
    else {
        ret.from = 0;
    }
    if (range.hasOwnProperty('to')) {
        const temp = utils.fromETHBlockNumber(range.to);
        if (temp !== 'best') {
            ret.to = temp;
        }
        else {
            ret.to = Number.MAX_SAFE_INTEGER;
        }
    }
    else {
        ret.to = Number.MAX_SAFE_INTEGER;
    }
    return ret;
};
exports.formatOptions = function (options) {
    const ret = {};
    if (options.hasOwnProperty('limit')) {
        const temp = utils.toInteger(options.limit);
        if (temp) {
            ret.limit = temp;
        }
    }
    if (options.hasOwnProperty('offset')) {
        const temp = utils.toInteger(options.offset);
        if (temp) {
            ret.offset = temp;
        }
    }
    if (ret.hasOwnProperty('limit') || ret.hasOwnProperty('offset')) {
        return ret;
    }
    else {
        return null;
    }
};
exports.formatLogQuery = function (params) {
    const body = {
        topicSets: [],
    };
    if (params.range) {
        const ret = exports.formatRange(params.range);
        if (ret) {
            body.range = ret;
        }
    }
    if (params.options) {
        const ret = exports.formatOptions(params.options);
        if (ret) {
            body.options = ret;
        }
    }
    if (!body.range && (params.hasOwnProperty('fromBlock') || params.hasOwnProperty('toBlock'))) {
        body.range = {
            unit: 'block',
        };
        if (params.hasOwnProperty('fromBlock')) {
            body.range.from = params.fromBlock;
        }
        if (params.hasOwnProperty('toBlock')) {
            body.range.to = params.toBlock;
        }
        body.range = exports.formatRange(body.range);
    }
    body.topicSets = [];
    const topics = [];
    if (params.topics && params.topics.length) {
        for (let i = 0; i < params.topics.length; i++) {
            if (typeof params.topics[i] === 'string') {
                topics.push({
                    name: 'topic' + i,
                    array: [params.topics[i]],
                });
            }
            else if (utils.isArray(params.topics[i]) && params.topics[i].length) {
                topics.push({
                    name: 'topic' + i,
                    array: params.topics[i],
                });
            }
        }
    }
    const outputTopic = function (inputTopics, index, receiver, current) {
        if (index === inputTopics.length) {
            const o = {};
            Object.assign(o, current);
            receiver.push(o);
            return;
        }
        for (const item of inputTopics[index].array) {
            current[inputTopics[index].name] = item;
            outputTopic(inputTopics, index + 1, receiver, current);
        }
    };
    if (topics.length) {
        outputTopic(topics, 0, body.topicSets, {});
    }
    return body;
};
